### PR TITLE
Create Queue Implementation

### PR DESCRIPTION
**Pull request by aneeshsharma**
Implemented a POST route to `/api/queue` to create a new queue.

For the `QueueStore` interface
- Created a simple structure, `MongoDB` which stores pointer to the MongoDB client as well as pointers to the Queue and Token collections.
- Implemented `QueueStore` methods for this `MongoDB` structure
- Methods other than `CreateQueue` panic with a "Not Implemented" error message
- `CreateQueue` only considers `queue.QueueName` from the argument and assigns the rest of the values itself

The connection to MongoDB is made in `cmd/api/main.go`. The `MongoDB` instance is passed to `InitializeRoutes` as a `QueueStore` which then injects it into the `handler` package. 

Another method I found was to pass it down to a handler creator function. The handler creator function would take the `MongoDB` instance as an argument and return a function which would handle the endpoint. (Function Closures Ref: https://go.dev/tour/moretypes/25)

Example -
```Go
func CreateQueueHandler(store persistence.QueueStore) {
    return func(w http.ResponseWriter, r *http.Request) {
        ...
    }
}
```

Also, implemented a handler for the `/api/queue` endpoint
- The function `handler.Queue` handles all requests to `/api/queue/` and calls the corresponding functions for `GET`, `POST` etc. using a switch case

Please let me know if you have any suggestions for changes or improvements in any of these.

### CODE REVIEW

**Comment by aneeshsharma** [comment]
For managing the Environment variables locally, I am thinking of using [this](https://github.com/joho/godotenv). I have used something similar on my NodeJS applications. This will allow each of us to have a separate `.env` file for local development containing environment variables for `MONGO_URI` etc. (which would be in `.gitignore` since we don't want to commit that)

File: internal/datastore/mongo_store.go
Comment (daltonfury42) [inline internal/datastore/mongo_store.go]: Let us take the queue object that we get as argument, and then put it as in.

No business logic should come here. This file is just mongo code:Code to store the passed queue into MongoDB.

File: internal/datastore/mongo_store.go
Comment (daltonfury42) [inline internal/datastore/mongo_store.go]: Why did we store to a global variable here? Can we avoid it?

File: internal/datastore/mongo_store.go
Comment (daltonfury42) [inline internal/datastore/mongo_store.go]: Rename `ConnectMongoDB`  -> `NewMongoDB`

File: internal/handler/queue.go
Comment (daltonfury42) [inline internal/handler/queue.go]: Instead of passing down the store to the mux from main, and then setting it there, let just initialise here? 

```
const store persistence.QueueStore = datastore.NewMongoDB()
```

**Review by daltonfury42 — COMMENTED** [review]
Thanks for the PR, this is great. First pass of review. 

I am also new to golang, so please discuss on the comments if you feel the need to.

File: internal/datastore/mongo_store.go
Comment (daltonfury42) [reply internal/datastore/mongo_store.go]: Tomorrow we might store queue object in portgres, and we would write an implementation for postgres. We don't want to copy the logic there.

File: internal/datastore/mongo_store.go
Comment (aneeshsharma) [reply internal/datastore/mongo_store.go]: oh thanks for pointing that out
I accidentally used the same variable as the global
Will update

File: internal/datastore/mongo_store.go
Comment (aneeshsharma) [reply internal/datastore/mongo_store.go]: Makes sense I'll move this to the endpoint instead
But I think we will still have to remove `Id` from it so that MongoDB generates an ID on its own.

Comment (aneeshsharma) [reply internal/datastore/mongo_store.go]: I've moved the initialization to the create queue handler
But, I've assigned `queue.Id = ""` to make sure that the Id is generated by Mongo

File: internal/handler/queue.go
Comment (aneeshsharma) [reply internal/handler/queue.go]: Can we change this separately after this PR? Calling `NewMongoDB` requires a URI which needs to be initialized somewhere.

I was thinking of creating a global config structure with the help of [godotenv](https://github.com/joho/godotenv)

**Comment by aneeshsharma** [comment]
@daltonfury42 I've made some changes. Please take a look.

File: cmd/api/main.go
Comment (daltonfury42) [inline cmd/api/main.go]: Let's move this and the above lines into the function `NewMongoDB()`

Later we will find a good configuration tool and move the stuff into a config file, but for now we will just keep the interfaces and layers clean.

File: internal/handler/common.go
Comment (daltonfury42) [inline internal/handler/common.go]: common.go was a good idea. Let's do two things. 

1. Move `internal/handler/common.go` to `internal/datastore/store.go`.
2. Let's also initialise Store to the mongo client in this file.
3. Then import Store and use it in internal/handler/queue.go

Comment (daltonfury42) [reply internal/handler/common.go]: Rationale: Our `persistence` package would then expose the store to outside world, it's implementation and everything would stay inside it as private.

File: internal/handler/queue.go
Comment (daltonfury42) [reply internal/handler/queue.go]: Let's hardcode the URL in the function, we can then move to reading it from some config / env / what you have in mind, as a follow up PR. See below comment.

**Comment by maaverik** [comment]
@aneeshsharma @daltonfury42 could you please spend some time creating a design document for this repo? We have [one](https://github.com/SimplQ/simplQ-frontend/blob/master/simplq/readme.md) for the frontend repo which details the tools used and guidelines on what components should go where. 

In this repo's case, I feel that there aren't going to be many tools used, but a brief description of the different components, intended folder structure, and overall flow of data would be useful not just for us to be on the same page, but also for more experienced go programmers to guide us and contribute more easily. 

Once we have a decent design doc ready, we could also set up individual GitHub issues or maybe a checklist of things to do, because there aren't a lot of sync ups between contributors except through PRs right now. Discussing the design of components in issues worked well for the frontend repo, it helped new contributors get up to speed from looking at the discussions and allowed them a space to ask questions.

Give it a thought please.

**Comment by daltonfury42** [comment]
Yes, we really need this.

Unlike having a separate doc like we did for frontend, I am thinking of exploring doing documentation inline in code as comments, and then use some document generator tool. That way the documentation will get maintained, and will evolve as we develop. What do you think?

**Comment by maaverik** [comment]
Documenting within code is definitely a good idea, and I think [godoc](https://go.dev/blog/godoc) is the recommended tool for it. I do still feel that we need a high level design doc to give overall direction for the project, basically collate things discussed in earlier PRs and mention the main abstraction strategies we intend to follow like for data store interfaces, testing interfaces etc. Using inline documentation will complement it and keep things up to date, like you said.

**Comment by daltonfury42** [comment]
Ok, I will take this up. 

I'd try to minimise the things captured in the main doc, and would try to push as much as content into the inline documentation. I might be able to pull this off by writing documentation at package level, let me see if that works.

Found [this](https://elliotchance.medium.com/godoc-tips-tricks-cda6571549b) article to provide a preview of godoc capabilities.

**Comment by aneeshsharma** [comment]
Ok I'll try to include inline documentation when I write new parts.

> @aneeshsharma @daltonfury42 could you please spend some time creating a design document for this repo? We have [one](https://github.com/SimplQ/simplQ-frontend/blob/master/simplq/readme.md) for the frontend repo which details the tools used and guidelines on what components should go where.
> 
> In this repo's case, I feel that there aren't going to be many tools used, but a brief description of the different components, intended folder structure, and overall flow of data would be useful not just for us to be on the same page, but also for more experienced go programmers to guide us and contribute more easily.
> 
> Once we have a decent design doc ready, we could also set up individual GitHub issues or maybe a checklist of things to do, because there aren't a lot of sync ups between contributors except through PRs right now. Discussing the design of components in issues worked well for the frontend repo, it helped new contributors get up to speed from looking at the discussions and allowed them a space to ask questions.
> 
> Give it a thought please.

Also, @maaverik Can you file this as an issue?
This PR would merge soon so its good to keep track of this somewhere else.

File: internal/handler/common.go
Comment (aneeshsharma) [reply internal/handler/common.go]: Updated

**Comment by maaverik** [comment]
> This PR would merge soon so its good to keep track of this somewhere else.

Good point, I've created #8 to keep track of our discussion.

File: internal/mux/mux.go
Comment (daltonfury42) [inline internal/mux/mux.go]: We don't need to pass this variable anymore

File: cmd/api/main.go
Comment (daltonfury42) [inline cmd/api/main.go]: We don't need this anymore.

**Review by daltonfury42 — APPROVED** [review]

File: internal/datastore/mongo_store.go
Comment (daltonfury42) [inline internal/datastore/mongo_store.go]: nit: s/Connection/Connecting

File: internal/datastore/mongo_store.go
Comment (daltonfury42) [inline internal/datastore/mongo_store.go:1]: datastore and persistence packages are redundant. Let's move 
```internal/persistence/persistence.go -> internal/datastore/store.go```

**Comment by aneeshsharma** [comment]
I've moved `QueueStore` to `datastore` package
If this looks good we can merge the PR

File: internal/datastore/mongo_store.go
Comment (daltonfury42) [inline internal/datastore/mongo_store.go]: nit: remove unused variable.

**Review by daltonfury42 — APPROVED** [review]
Please go ahead. Once you've got one approval, you can merge once all open comments are resolved.

**Pull request merged**
